/**:
  ros__parameters:
    log_to_file: true
    log_debug: true
    log_enabled:
        # navigator
      - navigation
      #- navigation.graph_map_server
      #- navigation.command
    
        # tactic
      - tactic
      - tactic.pipeline
      #- tactic.eop
      - tactic.module
      #- tactic.module.live_mem_manager
      #- tactic.module.graph_mem_manager

      # path planner
      - path_planning
      - path_planning.cbit
      - path_planning.cbit_planner
      - cbit.path_planning
      #- cbit_planner.path_planning
      #- cbit.debug
      - mpc.solver
      - mpc.speed_scheduler
      #- mpc.cost_function
      #- mpc.cbit
      #- mpc.debug
      - cbit.control
      #- obstacle detection.cbit
      #- grizzly_controller_tests.cbit
      #- ouster
        
        # mission planner
      #- mission.server
      - mission.state_machine

        # pose graph
      #- pose_graph

        # lidar pipeline
      - lidar.pipeline
      #- lidar.aeva_converter_v2
      #- lidar.preprocessing
      - lidar.preprocessing_doppler
      - lidar.preprocessing_curvature
      - lidar.odometry_icp
      - lidar.odometry_doppler
      #- lidar.odometry_map_maintenance
      - lidar.vertex_test
      - lidar.localization_map_recall
      - lidar.localization_icp
      - lidar.localization_daicp
      #- lidar.intra_exp_merging
      #- lidar.dynamic_detection
      #- lidar.inter_exp_merging
      #- lidar.ground_extraction
      #- lidar.obstacle_detection
      #- lidar.terrain_assessment

    robot_frame: w200_0066_base_link
    env_info_topic: env_info
    lidar_frame: w200_0066_aeva_lidar
    lidar_topic: /aeva/sensor/point_cloud
    gyro_frame: w200_0066_aeva_imu
    gyro_topic: /aeva/sensor/imu
    gyro_bias:
      x: -0.0115
      y: -0.0071
      z: -0.0049
    queue_size: 1
    graph_projection:
      origin_lat: 43.742778
      origin_lng: -79.465556
      origin_theta: 1.3
      scale: 1.0
    tactic:
      enable_parallelization: true
      preprocessing_skippable: false
      odometry_mapping_skippable: false
      localization_skippable: true
      task_queue_num_threads: 1
      task_queue_size: -1

      route_completion_translation_threshold: 0.2
      route_completion_angle_threshold: 0.1

      chain:
        min_cusp_distance: 0.5
        angle_weight: 1.0
        search_depth: 5
        search_back_depth: 10
        distance_warning: 5.0
        alpha: 1.0
      save_odometry_result: true
      save_odometry_vel_result: true
      save_localization_result: true
      visualize: true
      rviz_loc_path_offset:
        - 0.0
        - 0.0
        - 0.0
    pipeline:
      type: lidar
      preprocessing:
        - conversion
        #- filtering
        - doppler_filtering
        - curvature_filtering
      odometry:
        #- icp
        - doppler
        - mapping
        - vertex_test
        #- intra_exp_merging
        #- dynamic_detection
        #- inter_exp_merging
        - memory
      localization:
        - recall
        #- icp
        - daicp
        #- safe_corridor
        #- change_detection_sync
        - memory
      submap_translation_threshold: 1.5
      submap_rotation_threshold: 30.0
    preprocessing:
      conversion:
        type: lidar.aeva_converter_v2
        visualize: true

      filtering:
        type: lidar.preprocessing
        num_threads: 12
        crop_range: 40.0                # range cropping in meters

        frame_voxel_size: 0.3           # grid subsampling voxel size

        vertical_angle_res: 0.00349066  # vertical angle resolution in radians
        polar_r_scale: 2.0              # polar_r_scale x vertical_angle_res = nearest neighbor search radius for normal computation
        r_scale: 4.0                    # scale down point range by this value after taking log, whatever works
        h_scale: 2.0                    # scale down yaw(phi) by this value so that vertical resolution ~= horizontal resolution

        num_sample1: 10000              # max number of sample after filtering based on planarity
        min_norm_score1: 0.95           # min planarity score

        num_sample2: 10000              # max number of sample after filtering based on planarity
        min_norm_score2: 0.2            # 0.2 is when the incident angle 5/12 * pi
        min_normal_estimate_dist: 1.0   # minimum distance to estimate normal in meters
        max_normal_estimate_angle: 0.44 # must <1/2, this value will be timed by M_PI

        cluster_num_sample: 10000       # maxnumber of sample after removing isolated points

        visualize: true

      doppler_filtering:
        type: lidar.preprocessing_doppler
        azimuth_res: 0.00349066         # vertical angle resolution in radians
        azimuth_start: -0.994838        # -57 deg in rad         
        azimuth_end: 1.09956            #  63 deg in rad
        num_rows: 64                    # num of vertical scan lines
        num_cols: 601                   # 120 deg FOV, 0.2 deg resolution
        min_dist: 0.0                   # meters
        max_dist: 40.0                  # meters
        #
        root_path: "${WARTHOG}/ros2/aeriesii_ros2/sensor_config"
        downsample_steps: 2             # after image projection 
        bias_input_feat: ["range", "intensity", "medianrv"]
        var_input_feat: ["intensity", "medianrv", "rv_var5"]
        bias_polyorder: 3
        var_polyorder: 2                # only one sensor

        visualize: true

      curvature_filtering:
        type: lidar.preprocessing_curvature
        num_threads: 12

        crop_range: 40.0            # range cropping in meters
        num_sample: 10000           # max number of points after range cropping
        k_neighbors: 30             # number of neighbors to compute curvature (keep <50)

        downsample_teach_map: false # whether to downsample the frames used in **map building**. in repeat, always downsample plane and feature separately 
                                    # if false, use feature_voxel_size for whole frame (default)
                                    # if true, downsample plane and features separately using plane_voxel_size and feature_voxel_size
        plane_voxel_size: 2.0       # voxel size for plane
        feature_voxel_size: 0.3     # voxel size for feature

        # clustering parameters
        t: 4.0
        d_prime: 2.0
        ground_plane_threshold: 0.5 # max curvature to be considered as ground
        min_points_threshold: 3    # minimum number of points to keep a cluster

        visualize: true

    odometry:
      icp:
        type: lidar.odometry_icp

        # continuous time estimation
        use_trajectory_estimation: false
        traj_num_extra_states: 0
        traj_qc_diag:
          - 1.0
          - 0.1
          - 0.1
          - 0.1
          - 0.1
          - 1.0
        num_threads: 8
        first_num_steps: 2
        initial_max_iter: 4
        initial_max_pairing_dist: 1.5
        initial_max_planar_dist: 1.0
        refined_max_iter: 50
        refined_max_pairing_dist: 1.0
        refined_max_planar_dist: 0.3
        averaging_num_steps: 2
        verbose: false
        max_iterations: 1
        min_matched_ratio: 0.0
        visualize: true
        gyro_cov: 1e-3

      doppler:
        type: lidar.odometry_doppler
        #
        num_threads: 12
        traj_qc_diag: [1.0, 0.001, 0.001, 0.001, 0.001, 1.0]
        #
        ransac_seed: 0
        ransac_gyro: true
        ransac_max_iter: 100
        ransac_threshold: 0.2
        prior_threshold: 0.5        # m/s
        ransac_min_range: 5.0       # m
        integration_steps: 100
        zero_vel_tol: 0.1
        gyro_cov: [1.43918122e-04,1.65332526e-04,4.68710201e-05]
        # prior
        P0inv: [1e4, 1e4, 1e4, 1e6, 1e6, 1e6]
        Qkinv: [124.097194, 852.430330017, 448.704299478, 1128.19639192, 1230.02908934, 11316.9459369]
        Qzinv: [0.0, 100.0, 100.0, 33271.2755284, 20032.1558168, 0.0]
        max_trans_vel_diff: 0.5

      mapping:
        type: lidar.odometry_map_maintenance_v2

        map_voxel_size: 0.3

        crop_range_front: 80.0
        back_over_front_ratio: 0.5
        point_life_time: 20.0
        visualize: true
      vertex_test:
        type: lidar.vertex_test

        max_translation: 0.30
        max_rotation: 10.0
      intra_exp_merging:
        type: lidar.intra_exp_merging_v2
        depth: 6.0

        map_voxel_size: 0.3

        crop_range_front: 10.0
        back_over_front_ratio: 0.5
        visualize: true
      dynamic_detection:
        type: lidar.dynamic_detection
        depth: 12.0

        horizontal_resolution: 0.0008726646
        vertical_resolution: 0.00698132
        max_num_observations: 2000
        min_num_observations: 4
        dynamic_threshold: 0.3
        visualize: true
      inter_exp_merging:
        type: "lidar.inter_exp_merging_v2"

        map_voxel_size: 0.3
        max_num_experiences: 128
        distance_threshold: 0.6
        planar_threshold: 0.2
        normal_threshold: 0.8
        dynamic_obs_threshold: 1
        visualize: true
      memory:
        type: live_mem_manager
        window_size: 5
    localization:
      recall:
        type: lidar.localization_map_recall
        map_version: pointmap
        visualize: true
      icp:
        type: lidar.localization_icp
        use_pose_prior: true
        num_threads: 8
        first_num_steps: 2
        initial_max_iter: 4
        initial_max_pairing_dist: 1.5
        initial_max_planar_dist: 1.0
        refined_max_iter: 50
        refined_max_pairing_dist: 1.0
        refined_max_planar_dist: 0.3
        averaging_num_steps: 2
        verbose: false
        max_iterations: 1
        min_matched_ratio: 0.45
        target_loc_time: 100.0
        calc_gy_bias: true
        calc_gy_bias_thresh: 10.0            # [s]
      daicp:
        type: lidar.localization_daicp
        # general
        num_threads: 8
        target_loc_time: 100.0                # [ms] lower down to 60.0
        # iteration control
        first_num_steps: 2
        initial_max_iter: 4
        # data association
        lambda: 1.0                           # weight for spatial vs curvature in correspondence search: 0 = all curvature, 1 = equal weight after normalization
        curvature_similarity_thresh: 120.0    # threshold on curvature difference for correspondence rejection
        initial_max_pairing_dist: 1.5
        initial_max_planar_dist: 1.0
        ### daicp
        max_gn_iter: 1
        # daicp - degeneracy threshold
        degeneracy_thresh: 2000.0              # we use condition number to set the threshold
        # daicp - range and bearing noise model
        sigma_d: 0.02                         # range noise std (2cm)
        sigma_az: 0.000523599                 # azimuth noise std (0.03 degrees)
        sigma_el: 0.000523599                 # elevation noise std (0.03 degrees)
        # prior fusion
        use_pfuison: true       #TODO remove  # whether to use prior fusion (always fuse prior, this param is not used)
        max_pfusion_iter: 2                   # max iterations for prior fusion
        use_L2_loss: false                    # whether to use L2 loss in prior fusion
        robust_loss: 0.1                      # parameter for robust loss function in prior fusion
        verbose: false
        # refinement
        refined_max_iter: 20                  # DAICP outer loop iterations
        refined_max_pairing_dist: 1.0
        refined_max_planar_dist: 0.3
        averaging_num_steps: 2
        # outlier rejection
        trans_outlier_thresh: 0.5            # threshold on translation outlier rejection
        rot_outlier_thresh: 0.1               # threshold on rotation outlier rejection
        min_matched_ratio: 0.45             
        # online gyro bias estimation
        calc_gy_bias: false
        calc_gy_bias_thresh: 10.0             # [s]

      safe_corridor:
        type: lidar.safe_corridor
        lookahead_distance: 5.0
        corridor_width: 3.5
        influence_distance: 1.0
        resolution: 0.25
        size_x: 16.0
        size_y: 8.0
        visualize: true
      change_detection_sync:
        type: lidar.change_detection_v3
        detection_range: 8.0
        search_radius: 0.25

        negprob_threshold: 0.015 # was 0.015 # -1.86 without prior, 0.015 with prior # Jordy: I found I needed to bump this up abit (approx 0.075+) to reduce false positives
        use_prior: true
        alpha0: 3.0
        beta0: 0.03
        use_support_filtering: true
        support_radius: 0.25
        support_variance: 0.05
        support_threshold: 2.5

        influence_distance: 0.6 # was 0.5 Jordy # Note that the total distance where grid cells have values > 0 is min dist + influence dist, not influence dist!
        minimum_distance: 0.9 # was 0.3 Jordy

        # cost map
        costmap_history_size: 15 # Keep between 3 and 20, used for temporal filtering
        resolution: 0.25
        size_x: 16.0
        size_y: 8.0
        visualize: true
      memory:
        type: graph_mem_manager
        vertex_life_span: 5
        window_size: 3
      obstacle_detection:
        type: lidar.obstacle_detection
        z_min: 0.5
        z_max: 2.0
        resolution: 0.6
        size_x: 40.0
        size_y: 20.0
        run_async: true
        visualize: true
      ground_extraction:
        type: lidar.ground_extraction
        z_offset: 0.2
        alpha: 0.035
        tolerance: 0.25
        Tm: 0.3
        Tm_small: 0.1
        Tb: 0.5
        Trmse: 0.1
        Tdprev: 1.0
        rmin: 2.0
        num_bins_small: 30.0
        bin_size_small: 0.5
        num_bins_large: 10.0
        bin_size_large: 1.0
        resolution: 0.6
        size_x: 40.0
        size_y: 20.0
        run_async: true
        visualize: true
      terrain_assessment:
        type: lidar.terrain_assessment
        lookahead_distance: 15.0
        corridor_width: 1.0
        search_radius: 1.0
        resolution: 0.5
        size_x: 40.0
        size_y: 20.0
        run_online: false
        run_async: true
        visualize: true
    
    path_planning:
      type: cbit
      control_period: 100 # ms
      cbit:
        obstacle_avoidance: false
        obs_padding: 0.0
        curv_to_euclid_discretization: 10
        sliding_window_width: 12.0
        sliding_window_freespace_padding: 0.5
        corridor_resolution: 0.2
        state_update_freq: 2.0
        update_state: true
        rand_seed: 1

        # Planner Tuning Params
        initial_samples: 400
        batch_samples: 200
        pre_seed_resolution: 0.5
        alpha: 0.25
        q_max: 2.5
        frame_interval: 50
        eta: 1.0
        rad_m_exhange: 1.00
        initial_exp_rad: 0.75
        extrapolation: false
        incremental_plotting: false
        plotting: true
      costmap:
        costmap_filter_value: 0.01
        costmap_history: 15 # Note I dont think im actually using this one anymore, am using the one in the change detection v3 now

      speed_scheduler:
        planar_curv_weight: 0.0
        profile_curv_weight: 0.0 
        eop_weight: 1.0
        min_vel: 0.5
      mpc:
        # Controller Params
        extrapolate_robot_pose: false
        mpc_verbosity: false
        forward_vel: 1.0
        max_lin_vel: 1.5
        max_ang_vel: 0.5
        max_ang_acc: 5.0
        ang_alpha: 0.811
        lin_alpha: 0.0
        robot_linear_velocity_scale: 1.0 # Used to scale the output twist linear velocity messages by some constant factor to compensate internal low level control miscalibration
        robot_angular_velocity_scale: 1.0 # Used to scale the output twist angular velocity messages by some constant factor to compensate internal low level control miscalibration

        vehicle_model: "unicycle"

        # Misc
        command_history_length: 100